### ISSUE TITLE
Truncate must be based on a tombstone

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#11230]
Currently, the truncate operation uses a local, wall-clock time_point, as a barrier,
such that we discard sstables that have no updates after this time, and we keep a truncation record
to prevent resurrecting data before this time_point id the node restarts, when replaying the commitlog.

This is problematic for several reasons.
First and foremost, the distributed truncate operation, on all nodes is incoherent.
Each node uses a different time_point, based on when it handled the truncate rpc, so if the are writes in parallel to truncate, some nodes may truncate them while other nodes will not (basically requiring running repair after truncate if the application writes aren't quiesced before truncate).
Also, the logic for deciding about the authoritative replay position for truncate (on each node), is overly complicated at best and was prone to several errors and fixes - as demonstrated by the changes to the following assert over time: See https://github.com/scylladb/scylladb/blob/7f80602b010a973dd24953cf39161699e7afcb42/replica/database.cc#L2385-L2391

The comprehensive fix to this issue is to realize that truncate as any DELETE operation over cql, should use a proper tombstone with one authoritative api::timestamp value that should be applied on all nodes such it will logically delete any data written up until this timepoint, based on the eventual consistency tombstone logic.
Then compaction can kick-in and delete that data, again, as it does with any other tombstone, whether it be a partition, row, or range tombstone.  This can be done very efficiently (e.g. by extending cleanup compaction), by:
- detecting fully-deleted sstables by their max_timestamp <= truncate_tombstone.timestamp and deleting them as a whole without bothering to compact them
- ignoring sstables with min_timestamp > truncate_tombstone.timestamp
- and compacting the rest (using rewrite_sstables compaction), keeping only the live mutation fragments.

The truncate tombstone itself can propagate via the schema (and kept persistently in a system schema table) and be applied on each mutation_partition when it is create using the schema or upgraded to use the schema with truncate_tombstone.

The trickier parts with that would be:
- auto_snapshot the table (including flushing) before the truncate_tombstone becomes visible in the node
- purging the truncate_tombstone after gc_grace_period or after successful repair (in repair-mode gc).

**Comment by avikivity** [comment]
Note, truncate is not a critical operation. The usual way it is used is that the workload is acquiesced and then a truncate is issued.

**Comment by fee-mendes** [comment]
I think I stumbled upon the idea of a truncate tombstone once and if my mind is not failing me such concept would allow the TRUNCATE operation to be carried out without requiring a CL=ALL isn't it? Of course by lowering the consistency level requirements, one has to be careful to ensure that subsequent reads consistency properly avoid reading nodes that missed the given tombstone.

In any case, while I agree with @avikivity that TRUNCATE is not really a critical operation, it does have several design flaws such as the requirement for a CL=ALL, the fact that it takes absurdly longer on large tables and bigger deployments with many shards, and also the fact that it is not uncommon to see a scenario on when a table truncation creates noise to a workload running in another table in the same cluster.

**Comment by bhalevy** [comment]
It will stil require effectively CL=ALL since it applies to all tokens and hence it involves all replicas.
Using a truncate tombstone that's part of the schema will give it the performance and serialization characteristic of ALTER TABLE. And the actual removal of sstables (that is O(table data size)) will be done in the background, in maintenance priority class.

**Comment by avikivity** [comment]
It doesn't need CL=ALL, just QUORUM. If a replica misses the truncate tombstone then reconciliation will apply it there retroactively.

**Comment by bhalevy** [comment]
I think it's not exactly a quorum but rather ALL - QUORUM^, i.e all nodes but `RF - QUORUM(RF)`, i.e. less than or equal to half the QUOROM.

**Comment by bhalevy** [comment]
For example, for a 10 nodes cluster and table with RF=3, you need to truncate on 10-1 = 9 nodes, since you want any 2 (CL=QUORUM) reads to include the truncate tombstone.

**Comment by avikivity** [comment]
It's a QUORUM for every replica set (since TRUNCATE touches every replica set). It's similar to a QUORUM full scan SELECT that touches every replica set quorum. In practice it's what you wrote, though on an extremely large cluster we can tolerate two nodes down if they don't both participate in a any replica sets.

**Comment by michoecho** [comment]
Note: when experimenting with schema-based truncate (i.e. storing a table-wide tombstone in the schema), I realized that it won't easily work with `system` keyspace, because its schemas are hardcoded, not loaded from `system_schema.tables`. (Am I wrong)? So some special treatment would be needed there.

**Comment by bhalevy** [comment]
> Note: when experimenting with schema-based truncate (i.e. storing a table-wide tombstone in the schema), I realized that it won't easily work with `system` keyspace, because its schemas are hardcoded, not loaded from `system_schema.tables`. (Am I wrong)? So some special treatment would be needed there.

I'm not aware of a need to support TRUNCATE for system tables at all, so I don't think it's a problem in practice.

Also, we decided to focus on native truncate for tablets for the time being.
When vnodes will go away we could probably close this issue as "not planned".

**Comment by bhalevy** [comment]
> > I'm not aware of a need to support TRUNCATE for system tables at all, so I don't think it's a problem in practice.
> 
> So I thought, but I ran into this because there are some raft tests which truncate system.topology.
> 
> But if it's not needed in the field, then indeed we could just not support TRUNCATE for `system`.
> 
> > Also, we decided to focus on native truncate for tablets for the time being.
> 
> Fair enough. (But it adds the burden of having two different TRUNCATE semantics in the program).

The plan is to introduce a `truncate` tablet transition type that will be similar to `cleanup` just leaving the empty-shell of the tablet replica behind. For table-wide truncate we would build a transition plan that will flush and take a snapshot of the table similar (if not identical) to how it's done today, then it will truncate all tablet replicas, synchronously deleting all data in them - not based on a tombstone.  So the semantics wouldn't be any different.  The difference would be in the way the process is controlled and serialized with other tablet transitions like migration, splitting/merging, etc.

**Comment by michoecho** [comment]
> > > I'm not aware of a need to support TRUNCATE for system tables at all, so I don't think it's a problem in practice.
> > 
> > 
> > So I thought, but I ran into this because there are some raft tests which truncate system.topology.
> > But if it's not needed in the field, then indeed we could just not support TRUNCATE for `system`.
> > > Also, we decided to focus on native truncate for tablets for the time being.
> > 
> > 
> > Fair enough. (But it adds the burden of having two different TRUNCATE semantics in the program).
> 
> The plan is to introduce a `truncate` tablet transition type that will be similar to `cleanup` just leaving the empty-shell of the tablet replica behind. For table-wide truncate we would build a transition plan that will flush and take a snapshot of the table similar (if not identical) to how it's done today, then it will truncate all tablet replicas, synchronously deleting all data in them - not based on a tombstone. So the semantics wouldn't be any different. The difference would be in the way the process is controlled and serialized with other tablet transitions like migration, splitting/merging, etc.

Oh, when did we pivot to this direction? Just a few days ago we were talking about tombstone-based TRUNCATE.

**Comment by mykaul** [comment]
It'd be great to have a design document for this, even if it's a short one, to ensure we are all on the same page.

**Comment by bhalevy** [comment]
> > > > I'm not aware of a need to support TRUNCATE for system tables at all, so I don't think it's a problem in practice.
> > > 
> > > 
> > > So I thought, but I ran into this because there are some raft tests which truncate system.topology.
> > > But if it's not needed in the field, then indeed we could just not support TRUNCATE for `system`.
> > > > Also, we decided to focus on native truncate for tablets for the time being.
> > > 
> > > 
> > > Fair enough. (But it adds the burden of having two different TRUNCATE semantics in the program).
> > 
> > 
> > The plan is to introduce a `truncate` tablet transition type that will be similar to `cleanup` just leaving the empty-shell of the tablet replica behind. For table-wide truncate we would build a transition plan that will flush and take a snapshot of the table similar (if not identical) to how it's done today, then it will truncate all tablet replicas, synchronously deleting all data in them - not based on a tombstone. So the semantics wouldn't be any different. The difference would be in the way the process is controlled and serialized with other tablet transitions like migration, splitting/merging, etc.
> 
> Oh, when did we pivot to this direction? Just a few days ago we were talking about tombstone-based TRUNCATE.

You managed to convince me in the engineering summit.

**Comment by bhalevy** [comment]
> It'd be great to have a design document for this, even if it's a short one, to ensure we are all on the same page.

Will do

**Comment by kbr-scylla** [comment]
Guys can you rename the issue and update the original post (perhaps leave the existing descirption, but prepend with "here's the new proposal, the one below is outdated") to reflect the latest state (you apparently no longer want it to be based on tombstone)?

OR create a new issue?

I'm confused. @kostja you created a new issue (https://github.com/scylladb/scylladb/issues/18794) but it looks like a duplicate of this one, except that this one is no longer about using a tombstone but doing something else (https://github.com/scylladb/scylladb/issues/11230#issuecomment-2076488188) so can somebody explain and introduce some order into this mess?

I need to understand what the tablet team needs and wants to do. If you no longer want a timestamp-based truncate but something else then I'll close https://github.com/scylladb/scylladb/issues/18794 @kostja

**Comment by bhalevy** [comment]
> Guys can you rename the issue and update the original post (perhaps leave the existing descirption, but prepend with "here's the new proposal, the one below is outdated") to reflect the latest state (you apparently no longer want it to be based on tombstone)?
> 
> OR create a new issue?
> 
> I'm confused. @kostja you created a new issue (#18794) but it looks like a duplicate of this one, except that this one is no longer about using a tombstone but doing something else ([#11230 (comment)](https://github.com/scylladb/scylladb/issues/11230#issuecomment-2076488188)) so can somebody explain and introduce some order into this mess?
> 
> I need to understand what the tablet team needs and wants to do. If you no longer want a timestamp-based truncate but something else then I'll close #18794 @kostja

@nodep is working for a tablet-based solution for https://github.com/scylladb/scylladb/issues/16411 that is simpler than current implementation of truncate for vnodes, by using tablet transitions to truncate all tablets, while serializing with tablet migration or split/merge, for example.  But it is still not based on a tombstone.

On the long term, as @avikivity mentioned in https://github.com/scylladb/scylladb/issues/16411#issuecomment-1873005060, we'd still prefer truncate to be based on a tombstone as it is a DML operation, not DDL.
Whether this timestamp should be coordinated somehow by using group0 (https://github.com/scylladb/scylladb/issues/18794) or would it just be picked up by the coordinator like other write timestamps and propagated by other means [1], it's still early to call.

[1] we considered adding a truncate tomebstone to the schema - as ugly as it sounds, since it is not a DDL operation, but it is convenient to consume by readers and compaction in particular, but it cannot be purged easily, if at  all.
